### PR TITLE
Avoid native eager router graph breaks under Dynamo

### LINKED ISSUE #186354: Dynamo graph break in native bmm_outer_product eager router due to _is_cow_tensor / call_boxed
### 🐛 Describe the bug

Hi PyTorch team, thanks for the work on the new native override path.

We hit a Dynamo graph break in a larger distributed `torch.compile` training workload on recent PyTorch main. The user code is a regular `torch.matmul`, but during tracing it enters the native `bmm_outer_product` eager router and breaks on the COW check:

```python
# torch/_native/ops/bmm_outer_product/triton_impl.py

a_is_cow = torch._C._is_cow_tensor(a)
b_is_cow = torch._C._is_cow_tensor(b)
```

The relevant user code is conceptually:

```python
querys_K = torch.matmul(torch.reshape(querys, [-1, nh, 1, att_emb_size]), K)
```

The workload has tensor/distributed wrappers around the model, so I suspect this path may differ from a tiny direct `torch.compile(torch.bmm)` repro. A direct minimal `torch.compile` script with outer-product-shaped `bmm` did not reproduce for me, but the larger workload consistently reports graph breaks at the native eager router.

### Error/log snippet

With `TORCH_LOGS=graph_breaks TORCHDYNAMO_VERBOSE=1`:

```text
Graph break in user code at torch/_native/ops/bmm_outer_product/triton_impl.py:35
Graph Break Reason: Encountered graph break when attempting to trace CALL
Attempted to call function marked as skipped
Explanation: Dynamo does not know how to trace the builtin `torch._C..._is_cow_tensor.`
Developer debug context: module: torch._C, qualname: ..._is_cow_tensor, skip reason: cannot determine source file for torch._C (likely a C extension or builtin)

Most recent call last:
  torch/_native/registry.py:923 in eager_router
    result = _dispatch(args, kwargs, swallow_cond_exceptions=False)
  torch/_native/registry.py:913 in _dispatch
    matched = cond(*args, **kwargs)
  torch/_native/ops/bmm_outer_product/triton_impl.py:35 in _bmm_outer_product_cond
    a_is_cow = torch._C._is_cow_tensor(a)
```

There is also a related fallback graph break when the eager router reaches the boxed fallback path:

```text
Graph break in user code at torch/_native/registry.py:925
Graph Break Reason: CALL_FUNCTION_EX
Unsupported method call
Explanation: Dynamo does not know how to trace method `call_boxed` of class `_SafeKernelFunction`

  torch/_native/registry.py:925 in eager_router
    return _fallback.call_boxed(keyset, *args, **kwargs)
```

This seems related to the comment in `torch/_native/registry.py` saying that compile/export routers run under FakeTensor, where predicates such as `_is_cow_tensor` may be undefined, so cond exceptions are swallowed and no-match returns `NotImplemented`. In this workload, however, Dynamo still appears to trace through `eager_router`, where `_is_cow_tensor` and `_fallback.call_boxed` are not trace-safe.

### Question

Is this expected behavior for the native `bmm_outer_product` override, or should the eager router avoid running these non-traceable checks while Dynamo is compiling?

Would a reasonable fix be one of the following?

1. Route native overrides through the compile-safe `compile_router` when Dynamo is tracing.
2. Make the eager router return a trace-safe fallback under `torch.compiler.is_compiling()` instead of calling `_fallback.call_boxed`.
3. Avoid querying `_is_cow_tensor` in the compile-time predicate and treat this as a no-match/default lowering case.

A local workaround seems to be disabling the bmm native override before compilation:

```python
import torch._native
from torch._native.registry import deregister_op_overrides

deregister_op_overrides(disable_op_symbols="bmm")
```

But it would be useful to know what the intended fix/design is.

### Versions

Observed on a local main build:

```text
Base PyTorch commit before local changes: ef8a54fb6d161bd81ef349c2adc283589deccb9f
CUDA: 13.1
Python: 3.11
```

Thanks!

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @amjames @jataylo @azahed98

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #186388

Fixes #186354

The native override eager router is installed as the real backend kernel for
aten overrides. Normally that path should evaluate eager-only predicates and
fall back through the captured boxed kernel, which avoids recursively
redispatching to the same override.

A torch.device context can make Dynamo inline DeviceContext.__torch_function__
and trace into the native bmm eager router. In that trace-time path, the router
was evaluating eager-only code such as torch._C._is_cow_tensor and the boxed
fallback, which produced graph breaks before Inductor could use the normal aten
path.

Use torch.compiler.is_dynamo_compiling() to detect the narrow case where Dynamo
is actively tracing the Python router and emit the resolved aten overload
instead. This is deliberately narrower than torch.compiler.is_compiling(): the
broader compile-session flag may be true while eager code is actually running,
and redispatching to aten in that case can recurse back into the same backend
router. Ordinary eager execution still uses the existing predicate and captured
boxed fallback path.

The alternative of using torch.compiler.is_compiling() was rejected because it
is true for the whole compile session and can recurse in real eager execution.
Changing the bmm predicate to avoid _is_cow_tensor would only fix one native
predicate and would leave boxed fallback graph breaks for other overrides.

Test Plan:
- python test/python_native/test_registry.py -k test_compile_session_flag_falls_through_without_recursion
- python test/dynamo/test_repros.py -k test_device_context_matmul_avoids_native_bmm_router_graph_break
- python test/python_native/test_registry.py
- python test/python_native/test_decomp.py
- lintrunner -a
- git diff --check

Benchmark Results:
- python CUDA bmm eager-router smoke benchmark, shape (64, 8, 1) x (64, 1, 128), 1000 iterations: 0.025610 s total, 25.610 us/call after the fix.

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and reran the two Dr. CI failures locally: `test_cow_input_nn_functional_bilinear_cuda_float32` and `test_cow_input___rmatmul___cuda_float32` both passed. The targeted Dynamo/native tests and `lintrunner -a` also passed, so the new head should rerun CI against the updated base.

**Review by drisspg — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and resubmitted it to rerun CI. The two current Dr. CI failures, `test_cow_input_linalg_matrix_power_cuda_float32` and `test_cow_input_nn_functional_bilinear_cuda_float32`, both pass locally on the rebased head, as do the targeted native-router/Dynamo tests and `lintrunner -a`.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main to refresh CI. The two current Dr. CI new failures, `test_cow_input_svd_lowrank_cuda_float32` and `test_cow_input_nn_functional_scaled_dot_product_attention_cuda_float32`, both pass locally on the rebased head, as do the targeted native-router/Dynamo regression tests and related python_native suites.

**Review by slayton58 — APPROVED** [review]

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main to refresh CI. The two Dr. CI new failures from the prior head, `test_cow_input_svd_lowrank_cuda_float32` and `test_cow_input_nn_functional_scaled_dot_product_attention_cuda_float32`, both pass locally on the rebased head, as do the targeted native-router/Dynamo regression tests, related python_native suites, `lintrunner -a`, and `git diff --check`.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and resubmitted it. The prior Dr. CI new failures were checked locally: test_cow_input_matmul_cuda_float32 passes, AOTInductorLoggingTest.test_shape_env_reuse passes when bypassing an unrelated local torchvision CUDA-version mismatch, and the shufflenet benchmark could not be reproduced locally because torchbenchmark is not installed in this checkout. The focused regression/native tests and lintrunner pass locally; new CI is now queued on the updated head.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main, resolving a test placement conflict in `test/dynamo/test_repros.py` by retaining both the new main CUDA repro and this PR device-context matmul regression test. The current Dr. CI new failures `test_cow_input_nn_functional_bilinear_cuda_float32` and `test_cow_input_matmul_cuda_float32` pass locally, as do the focused native-router/Dynamo tests, related python_native suites, `lintrunner -a`, and `git diff --check`; the `tf_efficientnet_b0` timm job could not be reproduced locally because the local torchvision install has a CUDA-version mismatch with this PyTorch build.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated this PR to address the Dr. CI `PYTORCH_TEST_WITH_INDUCTOR=1` COW failure for `TestCompositeComplianceCUDA.test_cow_input___rmatmul___cuda_float32`. The native eager-router Dynamo shortcut now preserves COW fallback semantics, Dynamo guards supported `_is_cow_tensor` checks instead of baking in unguarded COW state, and FakeTensor/export cases fail closed. Local checks passed for the failing Inductor COW test, the device-context graph-break repro, python_native registry/decomp tests, Dynamo trace-rule tests, `lintrunner -a`, and `git diff --check`.

**Comment by jansel** [comment]
My agent says I updated this PR to address Claude's COW review feedback and the Dr. CI lazy_clone_view failures. The COW guard now treats Python-key tensors as guard misses instead of raising, the FakeTensor hint and get_real_value ordering are fixed, and COW checks after _lazy_clone now graph-break instead of folding stale COW state, including aliasing-through-view cases. Local checks passed for the CPU/CUDA lazy_clone_view CI repros, the prior Inductor COW rmatmul failure, the native registry/decomp suites, the DeviceContext bmm graph-break repro, Dynamo trace-rule tests, lintrunner, and diff checks. New CI is queued on the updated head.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated this PR to address Claude’s latest COW review feedback. The COW-state-changing scan now detects function-form `torch._lazy_clone(...)` as well as method-form `_lazy_clone`, the regression test forces that function form into the FX graph, the DeviceContext repro comment no longer claims a source-less local can be guarded, and the exported COW guard expression now uses the same safe matcher as runtime guard evaluation. Local checks passed: `python test/python_native/test_registry.py -v`, the CUDA DeviceContext bmm repro, `test/python_native/test_decomp.py -v`, `test/dynamo/test_trace_rules.py -v`, `lintrunner -a`, and diff checks. New CI is queued on head `22b8c9104a6`.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated this PR to address the latest Claude review feedback. The Dynamo `_is_cow_tensor` handler now uses the same sentinel-aware `_try_is_cow_tensor` path as guard evaluation for real values, so Python-key tensor subclasses produce a registered graph break instead of leaking `RuntimeError` / `InternalTorchDynamoError`. I added a fullgraph wrapper-subclass regression and clarified why the `_lazy_clone` function-form test forces that call into the FX graph. Local checks passed for the full native registry suite, the CUDA DeviceContext bmm repro, native decomp, trace rules, `lintrunner -a`, `git diff --check`, and the prior Dr. CI `test_forward_ad_linalg_householder_product_cuda_float32` with the local runtime library path.

**Comment by jansel** [comment]
@claude review these changes

File: torch/__init__.py [hunk 1]
```diff
@@ -1412,6 +1412,18 @@ def sym_fresh_size(expr: "IntLikeType") -> "IntLikeType":
...
+
+def _is_cow_tensor(input: _Any) -> builtins.bool:
+    if _C._dispatch_keys(input).has(_C.DispatchKey.Python):
+        raise RuntimeError("_is_cow_tensor is not defined for Python tensor subclasses")
+    return _is_cow_tensor_impl(input)
+
+
+_C._is_cow_tensor = _is_cow_tensor  # pyrefly: ignore[missing-attribute]
+
+
```
Comment (jansel) [inline torch/__init__.py]: Agent, update the C++ version of thi rather than monkey-patching in python.

Comment (jansel) [reply torch/__init__.py]: My agent says I moved this out of the Python monkey-patch and into the C++ torch._C._is_cow_tensor pybind binding in torch/csrc/Module.cpp. The Python wrapper in torch/__init__.py is removed, and test/python_native/test_registry.py now has a direct wrapper-subclass regression for the binding. Local checks passed for the rebuilt registry suite, DLPack exchange API CPU/CUDA tests, the CUDA DeviceContext repro, the Dr. CI-reported composite compliance test, lintrunner, and diff checks.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main after the mergebot cherry-pick conflict in `torch/csrc/Module.cpp`. The conflict was resolved by keeping current main’s removal of the obsolete storage data-pointer helper bindings while retaining this PR’s C++ `_is_cow_tensor` Python-subclass rejection. Local checks passed for `test/python_native/test_registry.py`, the CUDA DeviceContext bmm repro, `test/python_native/test_decomp.py`, `test/dynamo/test_trace_rules.py`, `lintrunner -a`, and diff checks.

**Comment by jansel** [comment]
My agent says I rebased this PR past the autorevert and fixed the lint failures that caused it: the CUDA Dynamo repro now names backend="inductor" explicitly, and the COW handler uses is_fake_tensor() instead of a direct FakeTensor isinstance check. Local validation passed after rebuilding: the CUDA repro, test/python_native/test_registry.py -v, full lintrunner -a, and git diff --check.

**Comment by jansel** [comment]
@claude is this safe to reland
